src/tutorials/0008-introduction-to-libp2p/02.md

```diff
@@ -0,0 +1,46 @@
+---
+    title: "The current problem with Peer-to-Peer Protocols"
+    type: "multiple-choice"
+---
+
+IPFS is not the first attempt at creating a peer-to-peer protocol. Peer-to-peer concepts and implementations have been around for decades.
+A few examples are: Skype (originally developed as a peer-to-peer video call app), Gnutella, KaZaA, eMule, and BitTorrent.
+
+## Peer-to-peer state of the art
+
+One of the main problems with peer-to-peer connections is discoverability: how can two or more peers find each other and share information? These apps all had to solve this and more issues and each of them implemented their own solutions with different ideas and approaches.
+
+In fact, IPFS found inspiration on the current and past state of the art to try and improve its peer-to-peer system.
+There were a lot scientific papers from the academia world that provided ideas on how to solve some of these issues. Those researches had some preliminary results, so there had to be code implementations that could be used and adapted.
+But the truth is that those implementations were really hard to find, or couldn't be found at all. Some of the most common situations would be:
+
+- Lack of good documentation or none at all
+- Restrictive licensing or no license to be found
+- Very old code with the last update more than a decade ago
+- No point of contact (no maintainer available to reach)
+- Closed source (or product) or the source doesn't exist anymore
+- No specification
+- Implementation doesn't expose a friendly API
+- Tightly coupled with a specific use
+
+
+So in the end, every time peer-to-peer protocols needed to be used, its challenges needed to be solved again and again because there was no way to re-use those software packages that implemented peer-to-peer protocols.
+
+One would think that this was a problem of the past.
+The open-source community has evolved over the years and built a robust ecosystem that provides thousands of open-source software packages that include a thorough battery of test suites, good documentation, and friendly APIs.
+But unfortunately, no good peer-to-peer protocol implementation seems to have come out that solves all of the presented issues.
+
+For the most part, peer-to-peer protocol implementations make a lot of assumptions about the environments they run on that compromise the protocol's future, and they also turned out to be incredibly hard to upgrade or adapt to new environments.
```

Comment:
Good point.
Companies/Products like BitTorrent, IPFS, and Skype created their own implementations of peer-to-peer protocols.

A protocol might mean a conceptual definition on how something should work. But we also call it protocol to any of its implementations sometimes.

So in this case, I am referring interchangeably to the peer-to-peer protocols as if they were the products or the protocols they implemented, because that's the case we are making. That's exactly the issue here: Skype created their own protocol, BitTorrent as well, IPFS as well, as many others.

> if the underlying connection tool is the protocol, then I would understand why that underlying thing isn't reusable because it was designed for a very narrow context.

Exactly 👍

Reply:
Okay, so just a rewording to make that more clear. How about replacing the sentence on line 33 with: "Historically, companies that produced products like IPFS, Skype, or BitTorrent created their own protocols to support them. Those protocols made a lot of assumptions about the environments they'd run on and the needs they'd meet, making them incredibly hard to upgrade or adapt to new environments."